File: src/theShortestRoute/FindMinimumCost.java
package theShortestRoute;

import java.io.*;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class FindMinimumCost {
    static class Node {
        int node;
        int value;

        public Node(int node, int value) {
            this.node = node;
            this.value = value;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        List<List<Node>> graph = new ArrayList<>(n + 1);

        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            String[] sev = br.readLine().split(" ");
            int start = Integer.parseInt(sev[0]), end = Integer.parseInt(sev[1]), value = Integer.parseInt(sev[2]);

            graph.get(start).add(new Node(end, value));
        }

        String[] sten = br.readLine().split(" ");
        int start = Integer.parseInt(sten[0]), end = Integer.parseInt(sten[1]);

        int[] dist = new int[n + 1];

        for (int i = 0; i < n + 1; i++) {
            dist[i] = Integer.MAX_VALUE;
        }

        dist[start] = 0;
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(start, 0));

        while(!queue.isEmpty()){
            Node node = queue.remove();

            if(node.value > dist[node.node]) continue;

            List<Node> list = graph.get(node.node);

            for(Node cur : list){
                if(dist[cur.node] > dist[node.node] + cur.value){
                    dist[cur.node] = dist[node.node] + cur.value;
                    queue.add(new Node(cur.node, dist[cur.node]));
                }
            }
        }

        bw.write(String.valueOf(dist[end]));
        bw.flush();
        bw.close();
    }
}

File: JS/theShortestRoute/findMinimumCost.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("dev/stdin").toString().split("\n")
    : `5
8
1 2 2
1 3 3
1 4 1
1 5 10
2 4 2
3 4 1
3 5 1
4 5 3
1 5`.split("\n");

const n = parseInt(input[0]);
const m = parseInt(input[1]);

const graph = Array.from(Array(n + 1), () => []);

for (let i = 2; i < m + 2; i++) {
  const [start, end, value] = input[i].split(" ").map(Number);

  graph[start].push({ node: end, value: value });
}

const [start, end] = input[m + 2].split(" ").map(Number);

const dist = Array(n + 1).fill(Infinity);

dist[start] = 0;
const queue = [{ node: start, value: 0 }];

while (queue.length > 0) {
  const { node, value } = queue.shift();

  if (value > dist[node]) continue;

  for (const { node: end, value: w } of graph[node]) {
    if (dist[end] > dist[node] + w) {
      dist[end] = dist[node] + w;
      queue.push({ node: end, value: dist[end] });
    }
  }
}

console.log(dist[end]);
